src/sage/algebras/lie_algebras/superliealgebra.py

```
@@ -0,0 +1,354 @@
+        cm = codomain.base_module()
+
+        spanning_set = [X.to_vector() for X in list(on_generators)]
+        im_gens = [Y.to_vector() for Y in list(on_generators.values())]
+
+        def solve_linear_system(A, b):
+            R = QQ
+            A_inv = A.solve_left(matrix.identity(A.ncols()))
+            M = A * A_inv
+            for Mi, bk in zip(M.rows(), b):
+                test_bk = sum((R(Mij) * bj for Mij,bj in zip(Mi,b)), cm.zero())
+                if test_bk != bk:
+                    raise ValueError("contradictory linear system")
+
+            return [sum((R(Aij) * bk for Aij,bk in zip(Ai,b)), cm.zero())
+                    for Ai in A_inv.rows()]
+
+        bracketlength = 1
+        n = 0
+        while True:
+            sm = m.submodule(spanning_set)
+            A = matrix(sm.base_ring(), [sm.coordinate_vector(X) for X in spanning_set])
+            try:
+                im_gens = solve_linear_system(A, im_gens)
+            except ValueError:
+                raise ValueError("this does not define a graded Lie algebra morphism")
+
+            spanning_set = list(sm.basis())
+            if n == len(spanning_set):
+
+                break
+
+            bracketlength += 1
+            n = len(spanning_set)
+            for i,j in combinations(range(n), 2):
+
+                Z = list(on_generators.keys())[i].bracket(list(on_generators.keys())[j])
+                imZ = list(on_generators.values())[i].bracket(list(on_generators.values())[j])
+                spanning_set.append(Z.to_vector())
+                im_gens.append(imZ.to_vector())
+
+        A = matrix(m.base_ring(), spanning_set)
+        im_gens = solve_linear_system(A, im_gens)
+
+        bh = lambda t: t
+        return cm.sum(bh(c) * im_gens[i] for i, c in (argument.to_vector()).items())
+
+
+    class Element(CombinatorialFreeModule.Element):
+         
+        #Returns the degree of the highest-degree monomial in any expression:
+        def degree(self):
```

This should raise an error if not homogeneous as the degree is not defined. However, the category should come with a mechanism so that you only need to implement a `degree_on_basis()` in the parent class, and then this method you get for free (i.e., no implementation necessary).